### ISSUE TITLE
老师您好，请问双卡使用run_sft_accelerate.sh进行sft，数据使用200万medical，max_train_samples=400000 ，per_device_train_batch_size=2， gradient_accumulation_steps=8，学习率=2e-5。为什么训练时候最大进度显示100000，但是训练到12500teps时候学习率就降到0不变了？第二次使用tochrun 使用run_sft.sh脚本，同样的设置，最大进度显示12500。

### ISSUE THREAD

**Issue by xrc666** [shibing624/MedicalGPT#489]
### Describe the Question
Please provide a clear and concise description of what the question is.

**Comment by shibing624** [comment]
我来帮您分析这个问题。先看一下相关脚本和训练代码。
您好！这是一个非常经典的 `accelerate launch` + HF `Trainer` 组合时 **梯度累积步数被重复计算** 导致的问题。简单说一句话：

> `accelerate launch` 和 HF `Trainer` 都会各自处理一次 `gradient_accumulation_steps`，两者叠加导致调度器和训练器的 "总步数" 计算偏差了 8 倍。

---

## 数字对不对得上

先验证一下您的数字：

- 样本数：400,000
- 总 batch size = `per_device_train_batch_size × num_gpus × gradient_accumulation_steps` = 2 × 2 × 8 = **32**
- 正确的总优化步数 = 400,000 / 32 = **12,500 步** ✅（这是 `run_sft.sh` 用 torchrun 跑出来的结果，是**对的**）

而 `run_sft_accelerate.sh` 显示 100,000 步 = 12,500 × 8，**恰好多了一个 `gradient_accumulation_steps` 的倍数**。

---

## 为什么 accelerate 会多算 8 倍

HuggingFace `Trainer` 内部有一段逻辑（`transformers/trainer.py` 中的 `_inner_training_loop`）：

```
num_update_steps_per_epoch = len(dataloader) // args.gradient_accumulation_steps
max_steps = num_update_steps_per_epoch * num_train_epochs
```

它假设 **dataloader 吐出来的是 micro-batch**，自己再除以 `gradient_accumulation_steps` 得到优化步数。

但当您用 `accelerate launch` 启动时，如果 accelerate 的配置文件里也设置了 `gradient_accumulation_steps`（或者 `Trainer` 内部把 accelerator 的 GA 也启用了），就会出现：

1. Trainer 看到 dataloader 长度 = 100,000（每个 GPU 50,000 step，2 卡 × per_device_batch 2 = 每 step 消费 4 条样本，400,000/4 = 100,000）
2. Trainer 该除以 8 得到 12,500，但因为 accelerate 那一层已经"吃掉"了 GA，Trainer 这里就没再除，于是 `max_steps = 100,000`
3. 学习率调度器按 `max_steps=100,000` 设置，warmup_steps=5 之后立刻线性衰减
4. 但**真正的优化 step**（`optimizer.step()` 被调用的次数）仍然只有 12,500 次
5. 所以走到第 12,500 个"真实优化 step"时，调度器的 internal counter 其实已经走到了 100,000（因为 lr_scheduler 每个 micro-batch 前向后都被 step 了一次），lr 就被衰减到 0 了

这就是您看到 "进度条显示 100,000，但 step 12,500 时 lr 就 = 0" 的根因。

而 `run_sft.sh` 用 **torchrun** 启动，走的是纯 `Trainer` 逻辑，没有 accelerate launcher 这一层，GA 只被除一次，所以 max_steps 正确显示 12,500。

---

## 解决办法（三选一）

### 方案 1（推荐）：不要用 accelerate launch 启动 Trainer

对于 HF `Trainer` 的脚本，**直接用 `torchrun`**（也就是您的 `run_sft.sh`），不要套 `accelerate launch`。Trainer 内部本来就集成了 accelerate，再套一层 launcher 反而重复处理 GA。

### 方案 2：检查/删除 accelerate config 里的 GA 设置

```bash
cat ~/.cache/huggingface/accelerate/default_config.yaml
```

把其中的 `gradient_accumulation_steps` 改成 1（或删掉），只让 Trainer 的 `--gradient_accumulation_steps 8` 生效。或者跑一次：

```bash
accelerate config
```

在 "How many gradient accumulation steps" 那一问选 **1**。

### 方案 3：显式传 accelerate 的 GA=1

```bash
accelerate launch \
    --multi_gpu \
    --num_processes=2 \
    --gradient_accumulation_steps 1 \
    ...
```

强制让 launcher 层的 GA=1，把梯度累积完全交给 Trainer 处理。

---

## 验证方法

改完之后重新启动，进度条最大值应该变成 **12,500**，和 `run_sft.sh` 一致，学习率按 12,500 步做 cosine/linear 衰减，训练末尾才接近 0。
